### PR TITLE
Make DictionaryVector::addNulls ensure indices_ size equal to vector size

### PR DESCRIPTION
**Pull request by kgpai**
Also set the newly added null indices point to 0th index.

### CODE REVIEW

**Comment by mbasmanova** [comment]
@kgpai Krishna, would you squash commits?

File: velox/vector/DictionaryVector.h [hunk 1]
```diff
@@ -176,6 +176,17 @@ class DictionaryVector : public SimpleVector<T> {
   void addNulls(const uint64_t* bits, const SelectivityVector& rows) override {
     flatNullsBuffer_ = nullptr;
     BaseVector::addNulls(bits, rows);
+
+    // Resize indices_ to be as large as nullsBuffer.
```
Comment (mbasmanova) [inline velox/vector/DictionaryVector.h:180]: typo: nullsBuffer -> nulls buffer

File: velox/vector/DictionaryVector.h [hunk 2]
```diff
@@ -176,6 +176,17 @@ class DictionaryVector : public SimpleVector<T> {
...
+    auto bufferPtr = mutableIndices(rows.end());
```
Comment (mbasmanova) [inline velox/vector/DictionaryVector.h:181]: Are we guaranteed that rows.end() >= length_? Probably not. Consider taking a max(rows.end(), length_).

naming: bufferPtr -> indices

File: velox/vector/DictionaryVector.h [hunk 3]
```diff
@@ -176,6 +176,17 @@ class DictionaryVector : public SimpleVector<T> {
...
+    auto rawIndices = bufferPtr->template asMutable<vector_size_t>();
```
Comment (mbasmanova) [inline velox/vector/DictionaryVector.h:182]: Is "template" needed here?

File: velox/vector/tests/VectorTest.cpp [hunk 4]
```diff
@@ -1671,3 +1671,35 @@ TEST_F(VectorTest, multipleDictionariesOverLazy) {
     ASSERT_EQ(i, dict->as<SimpleVector<int32_t>>()->valueAt(i));
   }
 }
+
+TEST_F(VectorTest, dictionaryAddNull) {
```
Comment (mbasmanova) [inline velox/vector/tests/VectorTest.cpp:1675]: naming: dictionaryAddNull -> xxxAddNulls

File: velox/vector/tests/VectorTest.cpp [hunk 5]
```diff
@@ -1671,3 +1671,35 @@ TEST_F(VectorTest, multipleDictionariesOverLazy) {
...
+  auto vectorMaker = std::make_unique<test::VectorMaker>(pool_.get());
```
Comment (mbasmanova) [inline velox/vector/tests/VectorTest.cpp:1676]: No need for a smart pointer here:

```
test::VectorMaker vectorMaker(pool_.get());
```

**Comment by kgpai** [comment]
@mbasmanova Rebased and squashed. Hope thats better.

File: velox/vector/tests/VectorTest.cpp [hunk 6]
```diff
@@ -1671,3 +1671,35 @@ TEST_F(VectorTest, multipleDictionariesOverLazy) {
...
+  auto dictionarySize = flatVector->size();
+  auto indices = allocateIndices(dictionarySize, pool_.get());
+  auto rawIndices = indices->asMutable<vector_size_t>();
+  for (auto i = 0; i < dictionarySize; i++) {
+    rawIndices[i] = i;
+  }
+  auto dictionaryVector = BaseVector::wrapInDictionary(
+      nullptr, indices, dictionarySize, flatVector);
+
+  // Try and append nulls to end of dictionary array.
```
Comment (mbasmanova) [inline velox/vector/tests/VectorTest.cpp:1688]: typo: to end -> at the end

typo: dictionary array -> dictionary vector

File: velox/vector/tests/VectorTest.cpp [hunk 7]
```diff
@@ -1671,3 +1671,35 @@ TEST_F(VectorTest, multipleDictionariesOverLazy) {
...
+  auto newDictionarySize = dictionarySize * 2;
+  SelectivityVector rows(newDictionarySize);
+
+  for (auto i = dictionarySize; i < newDictionarySize; i++) {
```
Comment (mbasmanova) [inline velox/vector/tests/VectorTest.cpp:1692]: rows.setValidRange for readability

**Review by mbasmanova — COMMENTED** [review]
@kgpai Krishna, thank you for making this change. Some comments below.

**Comment by mbasmanova** [comment]
@kgpai With this change, can we now remove the following code from Expr::addNulls?

```
  if (result->size() < rows.end()) {
    // Not all Vectors support resize.  Since we only want to append nulls,
    // we can work around that by calling setSize to resize the vector and
    // ensureNullsCapacity to resize the nulls_ bit vector.
    result->setSize(rows.end());
    result->ensureNullsCapacity(rows.end(), true);
  }
```

File: velox/vector/DictionaryVector.h (hunk 3 above)
Comment (kgpai) [reply velox/vector/DictionaryVector.h:182]: Yes, it doesnt compile without it.

**Comment by kgpai** [comment]
@mbasmanova  I dont think we can remove that as BaseVector::addNulls requires that length_ >= rows.end() and also we need to make nulls buffer as big. Let me know if I am missing something.

File: velox/vector/DictionaryVector.h (hunk 2 above)
Comment (kgpai) [reply velox/vector/DictionaryVector.h]: Thanks Masha, Good catch.

File: velox/vector/DictionaryVector.h [hunk 8]
```diff
@@ -176,6 +176,17 @@ class DictionaryVector : public SimpleVector<T> {
   void addNulls(const uint64_t* bits, const SelectivityVector& rows) override {
     flatNullsBuffer_ = nullptr;
     BaseVector::addNulls(bits, rows);
+
+    // Resize indices_ to be as large as nulls buffer.
```
[the remark this answers is not in the document]
Comment (kgpai) [reply velox/vector/DictionaryVector.h:180]: Reproducing offline discussion with Masha here and we should probably change all these functions : 

-> setSize()
-> append() 
-> resize()
-> copy() ? 
-> addNulls()
-> setNulls() //BaseVector has no check here on how big nulls is atm
-> resizeIndices() 
-> ensureBuffer()

In this PR I will focus only on addNulls.

**Pull request closed without merging**